README.md
# Cookiecutter Plone Starter 🌟

[![Cookiecutter Plone Project CI](https://github.com/collective/cookiecutter-plone-starter/actions/workflows/ci.yml/badge.svg)](https://github.com/collective/cookiecutter-plone-starter/actions/workflows/ci.yml)
[![Built with Cookiecutter](https://img.shields.io/badge/built%20with-Cookiecutter-ff69b4.svg?logo=cookiecutter)](https://github.com/collective/cookiecutter-plone-starter/)
![GitHub](https://img.shields.io/github/license/collective/cookiecutter-plone-starter)
[![Black code style](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/ambv/black)

Welcome to **Cookiecutter Plone Starter**! Your one-stop solution to kickstart [Plone](https://plone.org/) 6 projects with ease and efficiency. Powered by [Cookiecutter](https://github.com/cookiecutter/cookiecutter), this template is designed to save you time and ensure that you get started on the right foot. 🚀

## Features ✨

- Tailored for Plone 6
- Compatible with Python 3.8, 3.9, 3.10, 3.11
- Leverages [volto-generator](https://github.com/plone/volto/tree/master/packages/generator-volto) for frontend code generation

## Getting Started 🏁

### Prerequisites

- **pipx**: A handy tool for installing and running Python applications.
- **NodeJS & Yarn**: Essential for managing and running JavaScript packages.
- **Docker**: For containerization and easy deployment.

### Installation Guide 🛠️

1. **pipx**

```shell
pip install pipx
```

2. **NodeJS & Yarn**

Follow the [Plone documentation](https://6.docs.plone.org/volto/getting-started/install.html) for detailed instructions.

3. **Docker**

Visit the [official documentation](https://docs.docker.com/get-docker/) for installation guides.

### Generate Your Plone 6 Project 🎉

```shell
pipx run cookiecutter gh:collective/cookiecutter-plone-starter
```

For alpha versions of Volto:

```shell
USE_VOLTO_ALPHA=1 pipx run cookiecutter gh:collective/cookiecutter-plone-starter
```

### Initial Build

```shell
make install
```

### Start Servers

Backend:

```shell
make start-backend
```

Frontend:

```shell
make start-frontend
```

### Rebuild After Changes

```shell
make build
make start-backend
make start-frontend
```

## Project Generation Options 🛠️

Every project is unique, and we provide a variety of options to ensure that your Plone 6 project aligns with your specific needs. Here are the options you can customize during the generation process:

| Option                | Description                                                                                                                                          | Example                       |
| --------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------- |
| `project_title`       | Your project's human-readable name, capitals and spaces allowed.                                                                                     | **Plone Site**                |
| `description`         | Describes your project and gets used in places like ``README.md`` and such.                                                                          | **New site for our company.** |
| `project_slug`        | Your project's slug without spaces. Used to name your repository and Docker images.                                                                  | **plone-site**                |
| `hostname`            | Hostname where the project will be deployed.                                                                                                         | **site.plone.org**            |
| `author`              | This is you! The value goes into places like ``LICENSE``, ``setup.py`` and such.                                                                     | **Our Company**               |
| `email`               | The email address you want to identify yourself in the project.                                                                                      | **email@example.com**         |
| `python_package_name` | Name of the Python package used to configure your project. It needs to be Python-importable, so no dashes, spaces or special characters are allowed. | **plone_site**                |
| `volto_addon_name`    | Name of the Volto addon package used to configure your frontend project. No spaces or special characters are allowed.                                | **volto-plone-site**          |
| `python_test_framework` | Select which Python testing framework to use in the project                                                                                        | **pytest**                    |
| `plone_version`       | Plone version to be used. This queries for the latest available Plone 6 version and presents it to you as the default value.                         | **6.0.0**                     |
| `use_alpha_versions`  | Use non-stable versions of Volto, (The default value could also be set via `USE_VOLTO_ALPHA` environment variable.                                   | **Yes**                       |
| `volto_version`       | Volto (Plone Frontend) version to be used. This queries for the latest available Volto version and presents it to you as the default value.          | **16.4.1**                    |
| `volto_generator_version`       | Volto (Plone Frontend) Yeoman generator to be used. This queries for the latest available Volto Generator version and presents it to you as the default value. Non-stable versions will be included if a non-stable version was selected for `volto_version`.   | **6.0.0**                    |
| `language_code`       | Language to be used on the site.                                                                                                                     | **pt-br**                     |
| `github_organization` | Used for GitHub and Docker repositories.                                                                                                             | **collective**                |
| `container_registry`  | Container registry to be used.                                                                                                                       | **GitHub**                    |
| `devops_ansible`      | Should we create an Ansible playbook to bootstrap and deploy this project?                                                                           | **Yes**                       |
| `devops_gha_deploy`   | Should we create a GitHub action to deploy this project?                                                                                             | **Yes**                       |


## Dive into Your Project's Structure 🏗️

Your generated project will have a well-organized structure, ensuring that both development and maintenance are a breeze. It includes separate sections for backend, frontend, and devops, each tailored for its specific role.

(Include the Structure and Reasoning section from the previous README.md here, as it provides a good overview of the project structure)

## Code Quality Assurance 🧐

Your project comes equipped with linters to ensure code quality. Run the following to automatically format your code:

```shell
make format
```

## Internationalization 🌐

Generate translation files with ease:

```shell
make i18n
```

## License 📜

This project is licensed under the [MIT License](/LICENSE).

## Let's Get Building! 🚀

Happy coding!


{{ cookiecutter.project_slug }}/CHANGELOG.md
# Changes
## 1.0.0 ({{ cookiecutter.__generator_date_short }})

- Initial version [{{ cookiecutter.github_organization }}]


hooks/pre_gen_project.py
"""Pre generation hook."""
import re
import subprocess
import sys
from os import stat
from pathlib import Path
from textwrap import dedent
from typing import List

TERMINATOR = "\x1b[0m"
WARNING = "\x1b[1;33m"
INFO = "\x1b[1;34m"
HINT = "\x1b[3;35m"
SUCCESS = "\x1b[1;32m"
ERROR = "\x1b[1;31m"
MSG_DELIMITER = "=" * 80
SEMVER_PATTERN = r"^(?P<major>0|[1-9]\d*)\.(?P<minor>0|[1-9]\d*)\.(?P<patch>0|[1-9]\d*)(?:-(?P<prerelease>(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+(?P<buildmetadata>[0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$"  # noQA
PEP404_PATTERN = r"^(\d+!)?(\d+)(\.\d+)+([\.\-\_])?((a(lpha)?|b(eta)?|c|r(c|ev)?|pre(view)?)\d*)?(\.?(post|dev)\d*)?$"  # noQA


def _error(msg: str) -> str:
    """Format error message."""
    return f"{ERROR}{msg}{TERMINATOR}"


def _success(msg: str) -> str:
    """Format success message."""
    return f"{SUCCESS}{msg}{TERMINATOR}"


def _warning(msg: str) -> str:
    """Format warning message."""
    return f"{WARNING}{msg}{TERMINATOR}"


def _info(msg: str) -> str:
    """Format info message."""
    return f"{INFO}{msg}{TERMINATOR}"


SUPPORTED_PYTHON_VERSIONS = [
    "3.8",
    "3.9",
    "3.10",
    "3.11",
]


SUPPORTED_NODE_VERSION = [
    "16",
    "18",
    "20",
]


MIN_DOCKER_VERSION = "20.10"


output_path = Path().resolve()

context = {
    "project_title": "{{ cookiecutter.project_title }}",
    "project_slug": "{{ cookiecutter.project_slug }}",
    "description": "{{ cookiecutter.description }}",
    "author": "{{ cookiecutter.author }}",
    "email": "{{ cookiecutter.email }}",
    "python_package_name": "{{ cookiecutter.python_package_name }}",
    "volto_addon_name": "{{ cookiecutter.volto_addon_name }}",
    "plone_version": "{{ cookiecutter.plone_version }}",
    "volto_version": "{{ cookiecutter.volto_version }}",
    "language_code": "{{ cookiecutter.language_code }}",
    "github_organization": "{{ cookiecutter.github_organization }}",
    "container_registry": "{{ cookiecutter.container_registry }}",
}


def validate_not_empty(value: str) -> str:
    """Value should not be empty."""
    return "" if value.strip() else "should be provided"


def validate_plone_version(value: str) -> str:
    """Plone version should be superior to 6."""
    match = re.match(PEP404_PATTERN, value)
    valid = match and value.startswith("6.")
    return "" if valid else f"'{value}' is not a valid Plone version."


def validate_volto_version(value: str) -> str:
    """Volto version should be superior to 15."""
    match = re.match(SEMVER_PATTERN, value)
    valid = match and int(value.split(".")[0]) >= 15
    return "" if valid else f"'{value}' is not a valid Volto version."


def validate_language_code(value: str) -> str:
    """Language code should be valid."""
    pattern = r"^([a-z]{2}|[a-z]{2}-[a-z]{2})$"
    return (
        "" if re.match(pattern, value) else f"'{value}' is not a valid language code."
    )


def validate_python_package_name(value: str) -> str:
    """Validate python_package_name is an identifier."""
    return (
        "" if value.isidentifier() else f"'{value}' is not a valid Python identifier."
    )


VALIDATORS = {
    name: func for name, func in locals().items() if name.startswith("validate_")
}


def node_major_version(value: str) -> str:
    """Parse value and return the major version."""
    match = re.match(r"v(\d{1,3})\.\d{1,3}.\d{1,3}", value)
    return match.groups()[0] if match else ""


def docker_version(value: str) -> str:
    """Parse value and return the docker version."""
    value = value.strip()
    match = re.match(r"Docker version (\d{2}).(\d{1,2}).(\d{1,2})", value)
    if match:
        groups = match.groups()
        return f"{groups[0]}.{groups[1]}"
    return ""


def _get_command_version(cmd: str) -> str:
    """Get version of a command."""
    try:
        raw_version = (
            subprocess.run([cmd, "--version"], capture_output=True)
            .stdout.decode()
            .strip()
        )
    except FileNotFoundError:
        raw_version = ""
    return raw_version


def check_python_version() -> str:
    """Check if Python version is supported."""
    version = f"{sys.version_info.major}.{sys.version_info.minor}"
    return (
        ""
        if version in SUPPORTED_PYTHON_VERSIONS
        else f"Python version is not supported: Got {sys.version}"
    )


def check_node_version() -> str:
    """Check if node version is supported."""
    raw_version = _get_command_version("node")
    if not raw_version:
        return "NodeJS not found."
    else:
        version = node_major_version(raw_version)
        return (
            ""
            if version in SUPPORTED_NODE_VERSION
            else f"Node version is not supported: Got {raw_version}"
        )


def check_docker_version() -> str:
    """Check if docker version is supported."""
    raw_version = _get_command_version("docker")
    if not raw_version:
        return "Docker not found."
    else:
        version = docker_version(raw_version)
        return (
            ""
            if version >= MIN_DOCKER_VERSION
            else f"Docker version is not supported: Got {raw_version}"
        )


def check_yo_version() -> str:
    """Check if yo is installed."""
    raw_version = _get_command_version("yo")
    return "" if raw_version else "Yeoman not found."


def check_git_version() -> str:
    """Check if git is installed."""
    raw_version = _get_command_version("git")
    return "" if raw_version else "Git not found."


def sanity_check() -> bool:
    """Run sanity checks on the system."""
    checks = {
        "Python": {"func": check_python_version, "level": "error"},
        "Node": {"func": check_node_version, "level": "error"},
        "yo": {"func": check_yo_version, "level": "warning"},
        "Docker": {"func": check_docker_version, "level": "warning"},
        "git": {"func": check_git_version, "level": "warning"},
    }
    has_error = False
    print("Running sanity checks")
    for title, check_info in checks.items():
        func = check_info["func"]
        status = func()
        level = check_info["level"]
        if not status:
            msg = f"{_success('✓')}"
        elif level == "error":
            has_error = True
            msg = f"{_error(status)}"
        else:
            msg = f"{_warning(status)}"
        print(f"  - {title}: {msg}")
    return not (has_error)


def check_errors(data: dict) -> List[str]:
    """Check for errors in the provided data."""
    errors = []
    for key, value in data.items():
        func = VALIDATORS.get(f"validate_{key}", validate_not_empty)
        error = func(value)
        if error:
            errors.append(f"  - {key}: {_error(error)}")
    return errors


def main():
    """Validate context."""
    success = True
    value_errors = check_errors(context)
    print(f"{MSG_DELIMITER}")
    print(f"{ _info('{{ cookiecutter.project_title }} generation')}")
    print(f"{MSG_DELIMITER}")
    if not sanity_check():
        success = False
    elif value_errors:
        print("Value errors prevent running cookiecutter")
        for error in value_errors:
            print(error)
        success = False
    else:
        msg = dedent(
            f"""
            Summary:
              - Plone version: {_info('{{ cookiecutter.plone_version }}')}
              - Volto version: {_info('{{ cookiecutter.volto_version }}')}
              - Volto Generator version: {_info('{{ cookiecutter.volto_generator_version }}')}
              - Output folder: {_info(output_path)}
        """
        )

        print(msg)
    if not success:
        print(f"{MSG_DELIMITER}")
        sys.exit(1)


if __name__ == "__main__":
    main()
